keps/2941-DRA/README.md

```diff
@@ -0,0 +1,756 @@
+- This implementation focuses on structured parameters (beta in 1.32 and 1.33) and does not support alpha DRA features
+  like DRADeviceTaints, DRAAdminAccess, DRAPrioritizedLists, and DRAPartitionableDevices.
+
+### Risks and Mitigations
+
+With newer DRA features like DRAAdminAccess and DRAPrioritizedLists, there is a risk that
+effective tallying of resources will not be available until after allocation of these resources by kube-scheduler.
+
+In order to mitigate this risk, Kueue can take the following approach:
+1. For DRAPrioritizedLists: all the mentioned device classes in the request will be counted against the quota
+2. For DRAAdminAccess: This feature can only be enabled in admin namespace, therefore it should be skipped for being
+   counted against ClusterQuota. This is a different stance than Kubernetes ResourceQuota because kubernetes
+   ResourceQuota is namespace scoped. As a result, admin users can account for quota independent of user workloads.
+   On the contrary, with Kueue, since quota is part of ClusterQuota a cluster scoped object, admin workloads using
+   devices in admin namespace if counted against quota, will eat up quota meant for user workloads.
+3. For ResourceClaims with allocation mode `All`: worst-case scenario of the max number of devices that could be
+   allocated to a single claim will be used against quota.
+
+
+## Design Details
+
+A new feature gate DynamicResourceAllocation will be introduced, allowing users to test it in dev environments while
+making the changes dormant for production users. The following sections will explain the design in detail.
+
+### DynamicResourceAllocationConfig API
+
+```golang
+// DynamicResourceAllocationConfig is a singleton CRD that maps resource names todevice classes
+// used in ClusterQueue resource quotas. It is singleton as "default" is the only allowed named for the CRD instance in
+// Kueue namespace.
+type DynamicResourceAllocationConfig struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the desired state of DynamicResourceAllocationConfig
+    Spec DynamicResourceAllocationConfigSpec `json:"spec"`
+}
+
+// DynamicResourceAllocationConfigSpec defines the mappings between resource names and device classes
+type DynamicResourceAllocationConfigSpec struct {
+    // Resources is a list of mappings from resource name to device classes
+    // +listType=atomic
+    // +kubebuilder:validation:MaxItems=16
+    Resources []DynamicResource `json:"resources"`
+}
+
+// DynamicResource defines a mapping from a resource name to a list of device classes
+type DynamicResource struct {
+    // Name is the resource name that will be referred to in ClusterQueue and Workload status
+    Name corev1.ResourceName `json:"name"`
+
+    // DeviceClassNames lists the names of all the device classes that will count against
+    // the quota defined for this resource name
+    // +listType=atomic
```

Comment:
Same as another comment, do we need this as an atomic? Or is this field acceptable in any order?

Reply:
I'm ok with either typed, `map` / `set`. The point is if we can return the duplication errors to users when they accidentally add duplicated values.

@alaypatel07 Do you know if `set` supports raising validation error and return them to users? Or is it just output the error log in kueue-controller-manager?

Reply:
> > @alaypatel07 Do you know if set supports raising validation errors and returning them to users? Or is it just outputting the error log in the kueue-controller-manager?
> 
> I assumed this is going to be a openapi validation, so client will get the message through validation before it even hits the webhook/kueue stack

Thank you for describing that. In that case, we can select `set` instead of `map` which reduce annotations.

Reply:
> tbh I think like the validation can also be determined later during implementation phase, because then we will add a test, and now we rely static analysis.

It looks like the problems has been resolved. So, we can move to the next step.

Reply:
updated.